pilot/pkg/xds/ads.go

```diff
@@ -893,6 +917,8 @@ func (conn *Connection) send(res *discovery.DiscoveryResponse) error {
 	sendHandler := func() error {
 		start := time.Now()
 		defer func() { recordSendTime(time.Since(start)) }()
+		conn.streamMutex.Lock()
+		defer conn.streamMutex.Unlock()
```

Comment:
when does concurrent send happen? is it just for DebugType?

Reply:
right

Reply:
TBH I am quite concerned by this. We have had quite a lot of success serializing the request/push such that we don't need mutexes, etc everywhere. I am concerned that allowing so much concurrency from various places will cause a lot of complexity, if not bugs.

It seems the main benefit here is that we do not hold onto `discovery.Request` object as long, which I guess must be around 200kb in your tests. It seems like this is because `pushChannel` is triggered, blocking `reqChan`, causing it to retain the memory long? Is that accurate? Some heap pprofs would be handy.

If the issue is the channel blocking maybe we can prioritize the reqChan. Go doesn't have a perfect solution for this but this should be good enough: https://stackoverflow.com/a/11121616/374797. There is a small race but should be strictly better than today. Can you try that out and compare?

Reply:
We did the tests weeks ago, i could not distinguish which pprof belong to which case. 

But I have a simple report on the tests. Will send you

Reply:
@howardjohn The lock here can be removed, as the debug type is only from istioctl, it doesnot req concurrently with other xds